core/embed/trezorhal/stm32u5/secure_aes.c

```diff
@@ -58,19 +49,182 @@ static void secure_aes_load_bhk(void) {
+saes_invoke(void) {
+  // reset the key loaded in SAES
+  MODIFY_REG(SAES->CR, AES_CR_KEYSEL, CRYP_KEYSEL_NORMAL);
+
+  while (HAL_IS_BIT_SET(SAES->SR, CRYP_FLAG_BUSY)) {
+  }
+  while (HAL_IS_BIT_SET(SAES->ISR, CRYP_FLAG_RNGEIF)) {
+  }
+
+  MODIFY_REG(SAES->CR,
+             AES_CR_KMOD | AES_CR_DATATYPE | AES_CR_KEYSIZE | AES_CR_CHMOD |
+                 AES_CR_KEYSEL | AES_CR_KEYPROT,
+             CRYP_KEYMODE_NORMAL | CRYP_NO_SWAP | CRYP_KEYSIZE_256B |
+                 CRYP_AES_ECB | CRYP_KEYSEL_HSW | CRYP_KEYPROT_DISABLE);
+
+  TAMP->BKP0R;
+  TAMP->BKP1R;
+  TAMP->BKP2R;
+  TAMP->BKP3R;
+  TAMP->BKP4R;
+  TAMP->BKP5R;
+  TAMP->BKP6R;
+  TAMP->BKP7R;
+
+#define CRYP_OPERATINGMODE_ENCRYPT 0x00000000U /*!< Encryption mode(Mode 1) */
+
+  /* Set the operating mode and normal key selection */
+  MODIFY_REG(SAES->CR, AES_CR_MODE | AES_CR_KMOD,
+             CRYP_OPERATINGMODE_ENCRYPT | CRYP_KEYMODE_NORMAL);
+
+  SAES->CR |= AES_CR_EN;
+
+  for (int j = 0; j < KERNEL_UNPRIVILEGED_SIZE / AES_BLOCK_SIZE; j++) {
+    /* Write the input block in the IN FIFO */
+    SAES->DINR = saes_input[j * 4 + 0];
+    SAES->DINR = saes_input[j * 4 + 1];
+    SAES->DINR = saes_input[j * 4 + 2];
+    SAES->DINR = saes_input[j * 4 + 3];
+
+    while (HAL_IS_BIT_CLR(SAES->ISR, AES_ISR_CCF)) {
+    }
+
+    /* Clear CCF Flag */
+    SET_BIT(SAES->ICR, CRYP_CLEAR_CCF);
+
+    /* Read the output block from the output FIFO */
+    for (int i = 0U; i < 4U; i++) {
+      saes_output[j * 4 + i] = SAES->DOUTR;
+    }
+  }
+
+  SAES->CR &= ~AES_CR_EN;
+
+  // reset the key loaded in SAES
+  MODIFY_REG(SAES->CR, AES_CR_KEYSEL, CRYP_KEYSEL_NORMAL);
+
+  register uint32_t r0 __asm__("r0") = sectrue;
+  __asm__ volatile("svc %[svid]\n"
+                   :
+                   : [svid] "i"(SVC_CALLBACK_RETURN), "r"(r0)
+                   : "memory");
+  return 0;
+}
+
+extern uint32_t sram_u_start;
+extern uint32_t sram_u_end;
+
+secbool unpriv_encrypt(const uint8_t* input, size_t size, uint8_t* output,
+                       secure_aes_keysel_t key) {
+  if (size != KERNEL_UNPRIVILEGED_SIZE) {
+    return secfalse;
+  }
+
+  if (key != SECURE_AES_KEY_XORK_SN) {
+    return secfalse;
+  }
+
+  NVIC_SetPriority(SVCall_IRQn, IRQ_PRI_HIGHEST);
+  uint32_t basepri = __get_BASEPRI();
+  __set_BASEPRI(1);
+
+  mpu_mode_t mpu_mode = mpu_reconfig(MPU_MODE_SAES);
+
+  for (uint32_t* m = &sram_u_start; m < &sram_u_end; m++) {
+    *m = 0;
+  }
+
+  for (int i = 0; i < KERNEL_UNPRIVILEGED_SIZE; i++) {
+    ((uint8_t*)saes_input)[i] = input[i];
+  }
+
+  SAES->CR |= AES_CR_KEYSEL_0;
+
+  __HAL_RCC_SAES_CLK_DISABLE();
+  __HAL_RCC_SAES_FORCE_RESET();
+  __HAL_RCC_SAES_RELEASE_RESET();
+  __HAL_RCC_SAES_CLK_ENABLE();
+
+  secbool retval = invoke_unpriv(saes_invoke);
+
+  __HAL_RCC_SAES_CLK_DISABLE();
+  __HAL_RCC_SAES_FORCE_RESET();
+  __HAL_RCC_SAES_RELEASE_RESET();
+  __HAL_RCC_SAES_CLK_ENABLE();
+
+  for (int i = 0; i < KERNEL_UNPRIVILEGED_SIZE; i++) {
```

Comment:
@TychoVrahe Why not to use `memset` & `memcpy`?

Reply:
see https://github.com/trezor/trezor-firmware/commit/30248cbd49fd9e8d0a8fc2d4ccff7e3b1622d9a9